cli/factory.rs

```diff
@@ -1052,3 +1066,230 @@ impl CliFactory {
     })
   }
 }
+
+#[derive(Debug, Copy, Clone)]
+pub struct SpecifierInfo {
+  /// Type check as an ES module.
+  pub check: bool,
+  /// Type check virtual modules from doc snippets. If this is set but `check`
+  /// is not, this may be a markdown file for example.
+  pub check_doc: bool,
+}
+
+struct WorkspaceFileContainerEntry {
+  specifiers: Vec<(ModuleSpecifier, SpecifierInfo)>,
+  doc_snippet_specifiers: Vec<ModuleSpecifier>,
+  factory: CliFactory,
+  worker_factory: Arc<CliMainWorkerFactory>,
+}
+
+impl WorkspaceFileContainerEntry {
+  fn checked_specifiers(&self) -> impl Iterator<Item = &ModuleSpecifier> {
+    self
+      .specifiers
+      .iter()
+      .filter_map(|(s, i)| i.check.then_some(s))
+      .chain(self.doc_snippet_specifiers.iter())
+  }
+}
+
+pub struct WorkspaceFileContainer {
+  entries: Vec<WorkspaceFileContainerEntry>,
+}
```

Comment:
I put all the behaviour in here, it stores specifier sets grouped by workspace directory. I used it to dedup as much as I could around the type-checking invocations (`deno check`, `deno test` and `deno bench`). It handles cross-config type-checking, and the specifiers can be requested grouped by workspace directory along with whatever dir-specific service struct is needed. `--watch` is also deduped and simplified for these commands.